### ISSUE TITLE
"Other" card zone?

### ISSUE THREAD

**Issue by lodici** [magarena/magarena#86]
As far as I can tell the vaguely named "other" zone is used to displays cards from your library which match particular criteria specified by card abilities (eg. "...sacrifice a creature to look through your library for a creature card..."). So in fact, if this is the case, "other" is actually a filtered view of your library. Are there any examples where the cards displayed in the "other" zone are not sourced from the library?

**Comment by ShawnieBoy** [comment]
Kind of a broad area really, I would also include any .cardlist scenarios. Would include graveyard, exile etc. Essentially any zone can require a 'choice' of cards from the total cards from that zone.

'choose one of the exiled cards'
'choose one of the revealed cards'
'choose a creature card from an opponent's graveyard'

The 'other' zone would cover most 'reveal' situations as well.

**Comment by lodici** [comment]
Thanks - I won't rename to something with "library" in it then. Must be something better than "Other" though, "Choose/select cards" may be.

**Comment by ShawnieBoy** [comment]
No problem - "Choose card/s" would make sense. Most of the time when you have a selection of cards, you are choosing one or more of them. The only other times would be the simple, 'Look at target players hand' just as a simple reveal. As we're not implementing those yet, could be something simple to change contextually later on.

**Comment by lodici** [comment]
Actually is it not the case that...

> 'choose one of the exiled cards' and 'choose a creature card from an opponent's graveyard'

would automatically display the appropriate zone and highlight the valid cards?

> 'choose one of the revealed cards'

this I am assuming is from the library which has no designated zone and so would display the cards in ``Other``.

So in fact is it not true that the ``Other`` zone is just for displaying a selection of cards from the Library?

**Comment by ShawnieBoy** [comment]
In that case, I'd guess yes.

I assume that would turn the 'other' zone as actually being a view of the library, fitting in with the choosing from the other zones (hand, graveyard, exile). There is no 'other' zone :smile:

**Comment by lodici** [comment]
> I assume that would turn the 'other' zone as actually being a view of the library, fitting in with the choosing from the other zones (hand, graveyard, exile).
Exactly.

**Comment by melvinzhang** [comment]
The "other" zone is for showing a given set of cards, unlike the other zones which show only cards from one zone. Currently, it is used to show a subset of cards from your library, revealed cards. Future uses include cards from your opponent's hand (eg Thoughtsieze) or cards from your opponent's library (eg Seek).

It could also be from the opponent's hand, or any revealed cards.

**Comment by lodici** [comment]
> The "other" zone is for showing a given set of cards, unlike the other zones which show all cards from one zone. Currently, it is used to show a subset of cards from your library, revealed cards. Future uses include cards from your opponent's hand (eg Thoughtsieze) or cards from your opponent's library (eg Seek).

As far as I can tell we now have zone buttons for all possible cases because we have a Library and Hand zone button for each player. This means the "Other" zone can be dropped which in turn means the small vertical toggle buttons can be hidden which in turns means the possibility of hiding the little zone indicator icons, decluttering the board and reclaiming some horizontal space. Agree?

**Comment by ShawnieBoy** [comment]
:+1:
That's the only minibar I'll be glad to see go.

**Comment by lodici** [comment]
Alas, for it is not just a filtered list of cards from the library. There are four ``MagicChoice`` classes which can activate this "other" zone - ``MagicFromCardFilterChoice``, ``MagicFromCardListChoice``,``MagicScryChoice`` and ``MagicTargetChoice``. They do this by passing the list of cards to choose from through the poorly named ``SwingGameController.showCards``. 

When I test using Bow of Nylea and activate the "...put up to four target cards from your graveyard on the bottom of your library in any order" (jarring, it should be "..onto the bottom...") it uses this "other" zone to actually display the cards in the graveyard.

This seems inconsistent to me. Why can it not show the cards, highlighted using the the normal graveyard zone as other choices do?

**Comment by melvinzhang** [comment]
The separate area is used as the library and graveyard may have lot of
cards but we only want to show a few of them. This allows us to see the
possible cards without then being overlapped with the rest of the cards.

**Comment by ShawnieBoy** [comment]
~~~I'm wondering if it's more to do with the amount of cards that would be displayed in regards to the library - the graveyard doesn't normally exceed the libarary count, barring mill-decks and long, drawn-out duels.~~~ Melvin beat me to it
I'm surprised `MagicTargetChoice` is in there though

**Comment by melvinzhang** [comment]
Good point, there is an inconsistency between selecting one card and
several cards in the graveyard. Selecting several cards is relatively new
and only supported by the FromCardList or FromCardFilter choice, selecting
one card uses MagicTargetChoice.

**Comment by melvinzhang** [comment]
If we can enhance the card zone to prioritize showing selectable cards with
minimal overlap with the other cards, there would be no need to show then
separately in a different zone.

**Comment by lodici** [comment]
Thanks for the clarification and it does make sense when you have to select one or two cards from 50 odd. But we could have the same issue if the graveyard and exile zones fill up - essentially you are just moving cards between zones so what is an issue with the library could apply to the other zones later in a game.

Also presumably (duh!) in a real game you would be able to browse your whole library which would have some strategic value I would have thought - something you can't currently do in Magarena.

As it stands the UI behaviour is incorrect, it actually selects and animates the library zone which if I play a card like Bow of Nylea could be confusing.

Would it not be more consistent and simpler to have a transient "choice" view that applies in all cases where the player is required to select a card, be it one or twenty.  It would be automatically activated on a user prompt and only show the cards that can be chosen. However the player may then decide, before choosing a card, to check all the cards in their hand, graveyard, exile and even library zones by clicking on the appropriate zone button. You can do this now (apart from the library) but it leads to issue #559. However with the choice view, it is always re-displayed when the mouse is moved back over the user prompt.

**Comment by melvinzhang** [comment]
It depends on how this transient view is shown, also this view would have to indicate where the cards are coming from. A problem with the special zone is there is no indication of where the cards are from, you have to read the choice dialog below the player panel.

Highlighting the cards in the zone they are normally shown is more consistent. For example, selecting a card to play from your hand is also a choice of selecting a card. Currently the user clicks on the card in hand directly.

I think showing the cards in the place they are found is more intuitive then using the virtual "other" zone or even a transient view.

**Comment by lodici** [comment]
> I think showing the cards in the place they are found is more intuitive then using the virtual "other" zone or even a transient view.

Yes but this still brings up the problem you yourself raised above - how to clearly show the cards where there are so many cards that overlapping occurs. For example which do you prefer... 
![screenshot00608](https://cloud.githubusercontent.com/assets/2443519/26052010/0bc7d2b8-395c-11e7-880d-ad58231d3371.jpg)
versus
![screenshot00609](https://cloud.githubusercontent.com/assets/2443519/26052021/14b72d24-395c-11e7-9f35-85663a802ee5.jpg)

**Comment by melvinzhang** [comment]
Yes, for this to work nicely, the card zone layout needs to be aware of highlights. The UI should already be since it can draw these green boxes. It just needs to show the cards in green without any overlap and squeeze the other non-highlighted cards more.

**Comment by lodici** [comment]
> It just needs to show the cards in green without any overlap and squeeze the other non-highlighted cards more.

I would like to see some mock ups. Personally I can't see this working if you have a lot of choices amongst a lot of cards. Potentially, I think it would look very ugly at the minimum resolution (see images above) with the non-selected cards overlapping to such an extent that you would end up with a couple of pixels to select the card - [] [] [] [] [] [] [] ||||||||||

**Comment by melvinzhang** [comment]
Good point, it may not always be possible to show the choices without overlap if there are too many.  The goal will be to show the choices with less overlap compared to the rest of the cards in the zone. 

Not too good to mock-ups but here is that it would look like when the number of choices could fit, so are shown without overlap.

![mock-up-squeeze](https://cloud.githubusercontent.com/assets/1620634/26053138/42623988-399a-11e7-9b48-63f60e3f66ba.jpg)

It doesn't solve the select one out of many cards case, it would be as it is currently. That can only be solved if there is some kind of scrollable view.

**Comment by melvinzhang** [comment]
The problem is the zone treats all the cards the same when computing how much of it to be shown (eg how much the next card can overlap it). I think highlighted cards should be treated specially, more of it should be shown compared to the other cards which are not selectable.

**Comment by lodici** [comment]
> I think highlighted cards should be treated specially, more of it should be shown compared to the other cards which are not selectable.

But then you could argue why show the non-selectable cards at all?

> It doesn't solve the select one out of many cards case, it would be as it is currently. That can only be solved if there is some kind of scrollable view.

Sorry, not sure I follow what you are getting at. One card or many, surely it would be better to have a single consistent view for choosing a card (whatever that may eventually be).

**Comment by melvinzhang** [comment]
Because they are in the zone, to be consistent with playing a card from hand which is the most common form of card selection. We show the whole hand although not all the cards is playable, it would be very jarring for the hand cards to changed from phase to phase.

Yes, it would be better to have a consistent way to choose a card. The most common case is playing a card from hand (e.g. casting a spell, playing a creature). So selecting a card in a different occasion should match that.

**Comment by lodici** [comment]
> Yes, it would be better to have a consistent way to choose a card. The most common case is playing a card from hand (e.g. casting a spell, playing a creature). So selecting a card in a different occasion should match that.

It may be consistent with the Hand but it has to be usable with a much larger number of cards as would be the case with choosing cards from the library. I guess we need to see how it would work with a normal 60 card library and a more extreme 100+ card library which means someone coming up with some new code to layout the cards.

**Comment by melvinzhang** [comment]
Agree, consider the effect "Play a creature card from your graveyard without paying its mana cost". The UI should look like playing a card from hand, except now you are playing it from the graveyard instead, so it should show all graveyard cards with just the creature cards highlighted.
